deepchem/models/wandblogger.py

```diff
@@ -78,67 +86,224 @@ def __init__(self,
+      W&B chart panel section to log under
     """
-    if self.save_run_history:
-      history = self.wandb_run.history._data
-      self.run_history = history
+    # Rename keys to the correct category
+    if location is not None:
+      if location in self.location_ids:
+        for key in list(data.keys()):
+          new_key = location + "/" + str(key)
+          data[new_key] = data.pop(key)
+      else:
+        self.location_ids.append(location)
+        for key in list(data.keys()):
+          new_key = location + "/" + str(key)
+          data[new_key] = data.pop(key)
+
+    # Log to W&B
     if self.wandb_run is not None:
-      self.wandb_run.finish()
+      self.wandb_run.log(data, step=step)
+
+  def on_fit_end(self, data: Dict):
+    """Called before the end of training.
+
+    Parameters
+    ----------
+    data: Dict
+      Training summary values to be logged
+    """
+    # Set summary
+    if self.wandb_run is not None:
+      if "global_step" in data:
+        self.wandb_run.summary["global_step"] = data["global_step"]
+      if "final_avg_loss" in data:
+        self.wandb_run.summary["final_avg_loss"] = data["final_avg_loss"]
+
+  def save_checkpoint(self,
+                      model_dir: str,
+                      dc_model: Model,
+                      checkpoint_name: str,
+                      value_name: str,
+                      value: numeric,
+                      max_checkpoints_to_track: int,
+                      checkpoint_on_min: bool,
+                      metadata: Optional[Dict] = None):
+    """Save model checkpoint.
 
-  def update_config(self, config_data):
-    """Updates the W&B configuration.
     Parameters
     ----------
-    config_data: dict
-      additional configuration data to add
+    model_dir: str
+      directory containing model checkpoints
+    dc_model: Model
+      DeepChem model object to be saved
+    checkpoint_name: str
+      name of the checkpoint
+    value_name: str
+      the name metric to checkpoint on
+    value: numeric
+      the value of the metric to checkpoint on
+    max_checkpoints_to_track: int
+      the maximum number of checkpoints to track. New checkpoint must have
+      score better than the [max_checkpoints_to_track]th best checkpoint
+      in order to be saved.
+    checkpoint_on_min:
+      if True, the best model is considered to be the one that minimizes the
+      value. If False, the best model is considered to be the one
+      that maximizes it.
+    metadata: Dict, optional(default None)
+      metadata to be save along with the checkpoint
     """
-    self.wandb_run.config.update(config_data)
+
+    # Only called once when first checkpoint is saved to create tracking record
+    if (checkpoint_name not in self.best_models):
+      # Set up to track top values of this checkpoint
+      self.best_models[checkpoint_name] = {}
+      self.best_models[checkpoint_name]["model_values"] = []
+
+    # Sort in order
```

Comment:
All of the following is repeating a lot of logic that really belongs in ValidationCallback.  `save_checkpoint()` should simply save a checkpoint and nothing more.

Reply:
Yup agreed, I realized ValidationCallback already had logic for when to save checkpoints. I've removed the logic from save_checkpoint()